Separate behavior files into standalone C files

someone2639 opened the pull request:
Scope:
 - convert all .inc.c files in `src/game/behaviors` to standalone .c files. (COMPLETE)
 - ~~merge the remaining parts of obj_behaviors, obj_behaviors_2, object_helpers, and behavior_actions in some way.~~ (_Waaaaay out of scope; postponed to a future PR_)
 - Create a global object_fields and object_constants header that only has shared params, and localize the relevant local fields and constants to their behavior files. (COMPLETE)
 - Move relevant global object functions that were in behavior files to their "correct" files
 - Move toad_message to its own behavior file, and rename mario_misc.c to mario_geo.c

Tracked to-do list before PR is ready to merge:
~~Create a list of behavior files that need more splitting or naming.~~ (N/A)
 - Clean up. (DONE)

Commit:
add global object fields file for later

Commit:
isolate all files that could be done automatically

Commit:
prelim: rename all files that werent changed

Commit:
make all behavior files (even unfinished ones) use global includes

Commit:
start getting the game to compile

Commit:
make all obj_behaviors_2 files nonstatic

Commit:
add is_mario_attacking to interaction.c

Commit:
All behaviors are in .c files, game compiles

tracking warnings

Commit:
merge obj_behaviors into one file

Commit:
more consolidation in obj_behaviors headers

Commit:
forward-declare struct Camrea

Commit:
start deleting useless prototypes in obj_behaviors.h

Commit:
purge obj_behaviors_2.h

axollyon commented:
i see we're making 3.0 the "even more hellish to upgrade to than 2.0" update

Commit:
purge obj_behaviors_2, remove all duplicate init/loop funcs from obj_behaviors

someone2639 commented:
encouraging 3.0 to be the starting point for a romhack rather than another upgrade point is

the point

axollyon commented:
watch me

Commit:
address all warnings

Commit:
remove old object_fields.h

Commit:
debug

someone2639 commented:
Tracking a rogue issue that causes the o->oDoorForwardRoom and o->oDoorBackwardRoom values to lock to `-1` when rooms are present, but I do not know how to reproduce this (alignment issue?)

someone2639 commented:
TODO: Files that include `debug.h` need to also include the relevant header that defines the `UNUSED` qualifier

gheskett commented:
Have you actually profiled this? If I've learned anything from Kaze, it's that this might potentially be an atrocity for object runtimes. Also you still need copy the object action enums.

arthurtilly commented:
how would this affect performance? it shouldnt affect the compiled rom too much, i was under the impression this PR was just to make programming easier

gheskett commented:
Separate object files for every single object is presumably bad for the cache.

arthurtilly commented:
im confused as to what you think the difference in the final rom will be (outside of minor amounts of padding)

gheskett commented:
All I know is the madman kaze emulator threw his entire audio library into 3 files and had some weird justification for it, maybe this is a question for him (still profile it though so we at least have a non-speculative answer).

arthurtilly commented:
the files themselves don't matter much, kaze was more concerned about the total size of the audio library fitting inside one cache block - it wouldnt make a difference how many C files there were. the only real difference individual files make for compilation (aside from static variables) is having the start of each file be 16 byte aligned which is not really enough to make a difference. i think this pr would also cause the order of them to change but that would also not really cause a noticeable effect

someone2639 commented:
fwiw i did profile on funny emulator (theres screenshots in one of the channels) and found no significant difference

The door bug is what im worried about though...

arthurtilly commented:
is cozies' PR relevant? the one about resetting the door room array

gheskett commented:
As far as cache is concerned anyway, emulator profiling is not a useful metric.

gheskett commented:
> is cozies' PR relevant? the one about resetting the door room array

That PR addresses a vanilla bug

someone2639 commented:
It's a similar bug but i don't think it solves the issue

Commit:
Merge 2.2

Commit:
a

Commit:
git subrepo clone https://github.com/devwizard64/libcart src/libcart

subrepo:
  subdir:   "src/libcart"
  merged:   "948358b8"
upstream:
  origin:   "https://github.com/devwizard64/libcart"
  branch:   "main"
  commit:   "948358b8"
git-subrepo:
  version:  "0.4.6"
  origin:   "???"
  commit:   "???"

Commit:
merge upstream

Commit:
merge upstream again

Commit:
libcart is not here yet

Commit:
debugprint

Commit:
include math_util.h

Commit:
add math_util to every file that was missing it...

Commit:
this tool is out of scope

someone2639 commented:
**The door bug is now fixed** (as well as every other thing that could've broken due to that issue). This leaves Scope Bullet Point 2 (merging object helper files) as the last TODO for this PR.

Commit:
clean out behavior_actions.c; move toad and spinning star to their own behavior files

Commit:
rename mario_misc and behavior_actions

Commit:
cleanup

Commit:
remove the TODO's too as they seem to be addressed or duplicates

Commit:
i thought i set this back to 0

Commit:
fix build issue

Commit:
merge upstream

Commit:
address last two TODO's

someone2639 commented:
Did another run-through to make sure everything was in order; I believe this is ready for review now

aglab2 commented on src/game/camera.h:
```diff
@@ -90,6 +90,13 @@
+ **/
+
+/* End Birds */
+#define /*0x104*/ oEndBirdCutsceneVars9PointX OBJECT_FIELD_F32(0x1F)
```
Why have it here?

someone2639 replied:
It was a remnant of the behaviors that were still in camera.c, but now I guess it can just go to global_object_fields

aglab2 commented on src/game/obj_behaviors.c:
```diff
@@ -717,59 +709,711 @@ UNUSED s32 debug_sequence_tracker(s16 debugInputSequence[]) {
+    return dialogResponse;
+}
+
+void obj_set_dist_from_home(f32 distFromHome) {
```
All small functions like this should be `static inline` in header files.

someone2639 replied:
I did enough function moving here to get rid of `obj_behaviors_2` and `object_helpers_2`, but anything further is so out of scope that i have another in progress branch trying to consolidate and split everything in a way that makes sense

aglab2 replied:
I agree, let's keep this PR majorly to splitting behaviors to C files first to ease reviewing process. Can consider minor perf stuff later.

Commit:
address feedback (camera.h obj fields moved to global_object_fields)

arthurtilly commented:
Is this ready to review?

someone2639 commented:
yes (but even if you approve the changes it looks like i have to fix some merge conflicts either way)

arthurtilly reviewed (changes requested):
Get fixing them then!

Commit:
merge upstream

Commit:
remove duplicate headers

Commit:
Fix hidden compile errors

someone2639 commented:
done!

arthurtilly commented on include/global_object_fields.h:
```diff
@@ -0,0 +1,429 @@
+#ifndef GLOBAL_OBJECT_FIELDS_H
+#define GLOBAL_OBJECT_FIELDS_H
+
+/**
```
i don't see any reason to rename this file to something longer

arthurtilly commented on include/global_object_fields.h:
```diff
@@ -0,0 +1,429 @@
+#define /*0x088*/ oBowserCamAct          OBJECT_FIELD_S32(0x00)
+#define /*0x0F4*/ oBowserStatus          OBJECT_FIELD_S32(0x1B)
+
+/* Hoot */
```
still a bunch of specific ones here... if this is because they're referenced by other behaviors seems like a better solution is necessary
